File: fast_scroll_with_alphabet1/app/src/main/java/com/example/listwithsidealphabet/MainActivity.java
package com.example.listwithsidealphabet;


import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;

// https://github.com/code-computerlove/FastScrollRecyclerView
public class MainActivity extends AppCompatActivity {
    private RecyclerView mRecyclerView;
    private RecyclerView.Adapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mRecyclerView = (RecyclerView) findViewById(R.id.my_recycler_view);

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        mRecyclerView.setHasFixedSize(true);

        // use a linear layout manager
        mLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLayoutManager);

        // specify an adapter (see also next example)
        ArrayList<String> myDataset = new ArrayList<String>();
        for(int i=0; i<26; i++) {
            for(int j=1; j<=10; j++) {
                myDataset.add(Character.toString((char)(65 + i)) + " " + j + " Row item");
            }
        }
        HashMap<String, Integer> mapIndex = calculateIndexesForName(myDataset);

        mAdapter = new MyAdapter(myDataset, mapIndex);
        mRecyclerView.setAdapter(mAdapter);
        FastScrollRecyclerViewItemDecoration decoration = new FastScrollRecyclerViewItemDecoration(this);
        mRecyclerView.addItemDecoration(decoration);
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
    }

    private HashMap<String, Integer> calculateIndexesForName(ArrayList<String> items){
        HashMap<String, Integer> mapIndex = new LinkedHashMap<String, Integer>();
        for (int i = 0; i<items.size(); i++){
            String name = items.get(i);
            String index = name.substring(0,1);
            index = index.toUpperCase();

            if (!mapIndex.containsKey(index)) {
                mapIndex.put(index, i);
            }
        }
        return mapIndex;
    }

}

File: fast_scroll_with_alphabet2/app/src/main/java/com/example/list_with_side_apphabet/MainActivity.kt
package com.example.list_with_side_apphabet

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val mLayoutManager = LinearLayoutManager(this)
        rv_list.setLayoutManager(mLayoutManager)
        val myDataset = ArrayList<String>()
        for (i in 0..25) {
            for (j in 1..10) {
                myDataset.add(Character.toString((65 + i).toChar()) + " $j Row item")
            }
        }
        var mAdapter = MyAdapter(myDataset)
        rv_list.setAdapter(mAdapter)
        rv_list.setItemAnimator(DefaultItemAnimator())

        fast_scroller.setRecyclerView(rv_list)

        val mAlphabetItems = ArrayList<AlphabetItem>()
        val strAlphabets = ArrayList<String>()
        for (i in 0 until myDataset.size) {
            val name = myDataset.get(i)
            if (name == null || name!!.trim({ it <= ' ' }).isEmpty())
                continue

            val word = name!!.substring(0, 1)
            if (!strAlphabets.contains(word)) {
                strAlphabets.add(word)
                mAlphabetItems.add(AlphabetItem(i, word, false))
            }
        }

        fast_scroller.setUpAlphabet(mAlphabetItems)

    }
}
